internal/controller/stages/regular_stages.go

```diff
@@ -0,0 +1,1981 @@
+
+	// Find the Stage.
+	stage := &kargoapi.Stage{}
+	if err := r.client.Get(ctx, req.NamespacedName, stage); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Safety check: do not reconcile Stages that are control flow Stages.
+	if stage.IsControlFlow() {
+		return ctrl.Result{}, nil
+	}
+
+	// Handle deletion of the Stage.
+	if !stage.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.handleDelete(ctx, stage)
+	}
+
+	// Ensure the Stage has a finalizer and requeue if it was added.
+	// The reason to requeue is to ensure that a possible deletion of the Stage
+	// directly after the finalizer was added is handled without delay.
+	if ok, err := kargoapi.EnsureFinalizer(ctx, r.client, stage); ok || err != nil {
+		return ctrl.Result{Requeue: ok}, err
+	}
+
+	// Reconcile the Stage.
+	logger.Debug("reconciling Stage")
+	newStatus, needsRequeue, reconcileErr := r.reconcile(ctx, stage, time.Now())
+	logger.Debug("done reconciling Stage")
+
+	// Record the current refresh token as having been handled.
+	if token, ok := kargoapi.RefreshAnnotationValue(stage.GetAnnotations()); ok {
+		newStatus.LastHandledRefresh = token
+	}
+
+	// Patch the status of the Stage.
+	if err := kubeclient.PatchStatus(ctx, r.client, stage, func(status *kargoapi.StageStatus) {
+		*status = newStatus
+	}); err != nil {
+		// Prioritize the reconcile error if it exists.
+		if reconcileErr != nil {
+			logger.Error(err, "failed to update Stage status after reconciliation error")
+			return ctrl.Result{}, reconcileErr
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to update Stage status: %w", err)
+	}
+
+	// Return the reconcile error if it exists.
+	if reconcileErr != nil {
+		return ctrl.Result{}, reconcileErr
+	}
+	// Immediate requeue if needed.
+	if needsRequeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+	// Otherwise, requeue after a delay.
+	// TODO: Make the requeue delay configurable.
+	return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+}
+
+func (r *RegularStagesReconciler) reconcile(
+	ctx context.Context,
+	stage *kargoapi.Stage,
+	startTime time.Time,
+) (kargoapi.StageStatus, bool, error) {
+	logger := logging.LoggerFromContext(ctx)
+	newStatus := *stage.Status.DeepCopy()
+
+	// Mark the Stage as reconciling.
+	conditions.Set(&newStatus, &metav1.Condition{
+		Type:               kargoapi.ConditionTypeReconciling,
+		Status:             metav1.ConditionTrue,
+		Reason:             "Reconciling",
+		ObservedGeneration: stage.Generation,
+	})
+
+	var requestRequeue bool
+	subReconcilers := []struct {
+		name      string
+		reconcile func() (kargoapi.StageStatus, error)
+	}{
+		{
+			name: "syncing Promotions",
+			reconcile: func() (kargoapi.StageStatus, error) {
+				status, hasPendingPromotions, err := r.syncPromotions(ctx, stage)
+				if err != nil {
+					err = fmt.Errorf("failed to sync Promotions: %w", err)
+				}
+				if status.CurrentPromotion == nil && hasPendingPromotions {
+					requestRequeue = true
+				}
```

Comment:
Is this actually resolved?

Reply:
You had marked it as such, but I couldn't see what, if anything changed that fixed it. Or was it not actually an issue?

Reply:
The bit that "undoes" this is: https://github.com/akuity/kargo/blob/9414bb72f1cc2ae2f25eba03a342a874456b925e/internal/controller/stages/regular_stages.go#L400-L406

The reason I put it there is because it will get you a more accurate result, as it knows about the most recent verification changes at that time.